RealmSwift/SwiftUI.swift

```diff
@@ -643,6 +647,364 @@ extension EnvironmentValues {
+///            case .progress(let progress):
+///                ProgressView(progress)
+///            }
+///         }
+///     }
+///
+/// This opened `realm` can be later injected to the view as an environment value which will be used by our property wrappers
+/// to populate the view with data from the opened realm
+///
+///     ListView()
+///        .environment(\.realm, realm)
+///
+@available(iOS 14.0, macOS 11.0, tvOS 14.0, watchOS 7.0, *)
+@propertyWrapper public struct AsyncOpen<Partition>: DynamicProperty where Partition: BSON {
+    @Environment(\.realmConfiguration) var configuration
+    @Environment(\.partitionValue) var partitionValue
+    @ObservedObject private var storage: ObservableAsyncOpenStorage
+
+    private func asyncOpen() {
+        storage.asyncOpen()
+            .sink { completion in
+                if case .failure(let error) = completion {
+                    self.storage.asyncOpenState = .error(error)
+                }
+            } receiveValue: { realm in
+                self.storage.asyncOpenState = .open(realm)
+            }.store(in: &storage.cancellables)
+    }
+
+    /**
+     A Publisher for `AsyncOpenState`, emits a state each time the asyncOpen state changes.
+     */
+    public var projectedValue: Published<AsyncOpenState>.Publisher {
+        return storage.$asyncOpenState
+    }
+
+    /// :nodoc:
+    public var wrappedValue: AsyncOpenState {
+        storage.asyncOpenState
+    }
+
+    /**
+     This will cancel any notification from the property wrapper states
+     */
+    public func cancel() {
+        storage.cancellables.forEach { $0.cancel() }
+        storage.cancellables = []
+    }
+
+    /**
+     Initialize the property wrapper
+     - parameter appId: The unique identifier of your Realm app, if empty or `nil` will try to retrieve latest singular cached app.
+     - parameter partitionValue: The `BSON` value the Realm is partitioned on.
+     - parameter configuration: The `Realm.Configuration` used when creating the Realm,
+                 user's sync configuration for the given partition value will be set as the `syncConfiguration`,
+                 if empty the configuration is set to the `defaultConfiguration`
+     - parameter timeout: The maximum number of milliseconds to allow for a connection to
+                 become fully established., if empty or `nil` no connection timeout is set.
+     */
+    public init(appId: String? = nil,
+                partitionValue: Partition,
+                configuration: Realm.Configuration = Realm.Configuration.defaultConfiguration,
+                timeout: UInt? = nil) {
+        var app: App
+        if let appId = appId {
+            app = App(id: appId)
+        } else if RLMApp.appIds().count == 1, // Check if there is a singular cached app
+            let cachedAppId = RLMApp.appIds().first as? String {
+            app = App(id: cachedAppId)
+        } else if RLMApp.appIds().count > 1 {
```

Comment:
This should call `RLMApp.appIds()` exactly once and then reuse the local value. Multiple calls to it could potentially produce different values if a new App is created on a different thread while this is happening, and while it appears that it'll currently work reasonably if that does happen, that's the sort of thing that can produce weird and confused bugs.